### PR TITLE
fix: Join function does not properly handle divergent schemas being j…

### PR DESCRIPTION
**Pull request by skartikey**
…oined together

When one of the input streams passed into join contains tables with different schemas, it causes join to fail. This is because join produces the schema of its final output based on the first table it finds in each of the two input streams it receives. It does this once at the beginning of the transformation. Then, while it's processing tables, if it finds a column that it doesn't recognize as part of the schema, it throws an error.

Ideally, the join transformation would be able to handle this situation gracefully by just adding the newly found column to the schema, and populating any rows that don't have a value for that column with nulls. The goal of this fix is to modify join so that it can do just that.

Related issues - https://github.com/influxdata/flux/pull/4310

fixes: https://github.com/influxdata/flux/issues/4506 https://github.com/influxdata/flux/issues/4315

### Done checklist
- [ ] docs/SPEC.md updated
- [ ] Test cases written

### CODE REVIEW

File: stdlib/universe/join.go [hunk 1]
```diff
@@ -853,8 +857,19 @@ func (c *MergeJoinCache) missingColIdx(lCols, rCols []flux.ColMeta) []int {
...
+			newColumnIdx, err = c.addTableColumnsToSchema(lCol, column, builder)
+			if err != nil {
+				return nil, err
+			}
+		} else {
+			newColumnIdx, ok = c.colIndex[newColumn]
+			if !ok {
+				errors.Newf(codes.Internal, "could not find index for column '%s'", column.col)
+			}
+			trackIdx[newColumnIdx] = struct{}{}
```
[the remark this answers is not in the document]
Comment (onelson) [reply stdlib/universe/join.go:871]: Got it! Thanks. It was just curious to me that the blocks weren't symmetrical :sweat_smile:.

**Review by onelson — APPROVED** [review]

**Review by wolffcm — COMMENTED** [review]
I think you said you are still going to add more tests here, is that right?

The product changes look good to me. This version of join has lots of strange edge cases and is hard to understand. Hopefully we can replace it with something just as useful but less difficult to reason about.

For the tests you will be adding, what additional cases did you have in mind? The cases that I think tend to be confusing and bug prone are those where some of the columns in the group key overlap with the join key, and/or where the group key is different between the two sides. Adding some cases like these that include missing columns/heterogeneous schemas seems like a good idea.

I can still think of some cases that should produce an error, like when a column exists on both sides but has a different type. I wonder if we have cases that cover that?

**Comment by skartikey** [comment]
Added following test cases today - 

1) same column on both the streams with different types. [Passed]
2) different group keys on the left and right stream join on the nongroup key. [Passed]
3) different group keys on the left and right stream join on the group key. [Failed]

In the third case, the code is only joining the very first table from the left and right stream and discarding the rest of the tables. I am looking into it.
